<a href="https://colab.research.google.com/github/rachlef/deep-learning-challenge/blob/main/deep_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Preprocessing

In [ ]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import numpy as np

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [ ]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
newapp_df=application_df.drop(columns=["EIN", "NAME"]) 
newapp_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [ ]:
# Determine the number of unique values in each column.
print(newapp_df.nunique())

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64


In [ ]:
# Look at APPLICATION_TYPE value counts for binning
app_type_counts= newapp_df["APPLICATION_TYPE"].value_counts()
app_type_counts
#print(newapp_df["APPLICATION_TYPE"].value_counts())

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
from tables import index
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace= list(app_type_counts[app_type_counts<500].index)

# Replace in dataframe
for app in application_types_to_replace:
    newapp_df['APPLICATION_TYPE'] = newapp_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
newapp_df['APPLICATION_TYPE'].value_counts()


T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [ ]:
# Look at CLASSIFICATION value counts for binning
class_counts=newapp_df["CLASSIFICATION"].value_counts()
class_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [ ]:
# You may find it helpful to look at CLASSIFICATION value counts >1
newapp_df["CLASSIFICATION"].value_counts().loc[lambda x:x >1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [ ]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace= list(class_counts[class_counts<1000].index)

# Replace in dataframe
for cls in classifications_to_replace:
    newapp_df['CLASSIFICATION'] = newapp_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
newapp_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [ ]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies_newapp_df= pd.get_dummies(newapp_df, dtype=float)
dummies_newapp_df.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                     0.0   
1       1   108590              1                     0.0   
2       1     5000              0                     0.0   
3       1     6692              1                     0.0   
4       1   142590              1                     0.0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                   1.0                   0.0                  0.0   
1                   0.0                   0.0                  1.0   
2                   0.0                   0.0                  0.0   
3                   0.0                   0.0                  1.0   
4                   0.0                   0.0                  1.0   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                  0.0                  0.0                  0.0  ...   
1                  0.0                  0.0                  0.0  ...   
2                  0.0                  1.0                  0.0  ...   
3                  0.0                  0.0                  0.0  ...   
4                  0.0                  0.0                  0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 44 columns]

In [ ]:
# Split our preprocessed data into our features and target arrays
#X= dummies_newapp_df.drop("IS_SUCCESSFUL", axis=1).values
#y= dummies_newapp_df["IS_SUCCESSFUL"].values
X = dummies_newapp_df.drop(['IS_SUCCESSFUL'], axis=1).values
y = dummies_newapp_df['IS_SUCCESSFUL'].values
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test= train_test_split(X, y, random_state=42)
s =StandardScaler()
X_train = s.fit_transform(X_train)
X_test = s.fit_transform(X_test)

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [ ]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len( X_train_scaled[0])
hidden_nodes_layer1= 42
hidden_nodes_layer2=28
hidden_nodes_layer3=1


nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation='relu'))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation='relu'))


# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 42)                1848      
                                                                 
 dense_4 (Dense)             (None, 28)                1204      
                                                                 
 dense_5 (Dense)             (None, 1)                 29        
                                                                 
Total params: 3,081
Trainable params: 3,081
Non-trainable params: 0
_________________________________________________________________


In [ ]:

# Compile the model
nn.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics=['accuracy'])


In [ ]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train, epochs= 126)

Epoch 1/126
804/804 [==============================] - 2s 1ms/step - loss: 0.5754 - accuracy: 0.7202
Epoch 2/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5572 - accuracy: 0.7287
Epoch 3/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5540 - accuracy: 0.7297
Epoch 4/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5518 - accuracy: 0.7294
Epoch 5/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5507 - accuracy: 0.7290
Epoch 6/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5496 - accuracy: 0.7313
Epoch 7/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5485 - accuracy: 0.7313
Epoch 8/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5476 - accuracy: 0.7308
Epoch 9/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5475 - accuracy: 0.7319
Epoch 10/126
804/804 [==============================] - 1s 1ms/step - loss: 0.5468 - accura

In [ ]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5620 - accuracy: 0.7286 - 294ms/epoch - 1ms/step
Loss: 0.5619533061981201, Accuracy: 0.7286297082901001


In [ ]:
# Export our model to HDF5 file
from google.colab import files

nn.save('/content/AlphabetSoupCharity.h5')
files.download('/content/AlphabetSoupCharity.h5')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Alphabet Soup Application Report**




**Overview:** What the purpose of this analysis?

The purpose of this analysis is to asses the accuracy and facilitate a prediction in opting for which type of applicants for funding with the best chance of success in their ventures via utlizing machine learning and neural networks; which will be successful if funded by the nonprofrit, Alphabet Soup. This is done through the usage of the provided dataset depicting reports on (columns of),EIN and NAME—Identification columns,APPLICATION_TYPE—Alphabet Soup application type,AFFILIATION—Affiliated sector of industry,CLASSIFICATION—Government organization classification,USE_CASE—Use case for funding,ORGANIZATION—Organization type,STATUS—Active status, INCOME_AMT—Income classification,SPECIAL_CONSIDERATIONS—Special considerations for application,ASK_AMT—Funding amount requested and,
IS_SUCCESSFUL—Was the money used effectively. 

**Results:**

*What variable(s) are the target(s) for your model?*



1.   The target variable for the model was "Its_Successful"


*What variable(s) are the features for your model?*



1.  "Application_type"
2.  "Affiliation"
3. "Classification"
4. "Use_case"
5. "Organization"
6. "Status"
7. "Income_amount"
8. "Special_consideration"
9. "Ask_amount"



*What variable(s) should be removed from the input data because they are neither targets nor features?*


There were two variables I removed from the original DF,
1.   "Name"
2.   "EIN"

*How many neurons, layers, and activation functions did you select for your neural network model, and why?*



1.   I opted for 3 layers due to the nature of the size of the data. For the input layer i chose 42 neurons because the number of neurons in the input layer should equate to the number of features aka columns in the data utilized. In this case, there are 42 columns in my "dummies_newapp_df".  
2.   In the second layer, I used 2/3 the amount of the input layer plus the output layer which provided 28 neurons.
3. For my output layer I utlized 1 for the number of neurons, because rule of thumb is that for a classifier its best to use 1 neuron when a softmax is not in place. 
4. I chose relu and sigmond as the activiation function due to the greediness of relu and for simoid because it exists between (0 to 1) and its chosen to predict the probability as an output.Since our target variable is a probability of anything exists only between the range of 0 and 1, sigmoid is the right choice for the outer layer. 


*Were you able to achieve the target model performance?*


1.  No, I was not able to reach the target model performance  of higher than 75%. My max reach was 73%. 




*What steps did you take in your attempts to increase model performance?*


1.   I altered amounts bins
2.   Number of hidden layers
3. Number of neurons
4. adusted the epoch frequency

**Summary:**

Refer to second notebook. 

